File: apps/api/app/schemas/transactions.py
"""Transaction schemas — client lists own payout ledger. Read-only.

No create schema: transactions are never created via this router — a future
money-layer producer inserts rows on a bypass session, the same mechanism as
services.notifications.emit_notification.
"""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel

from app.models.transaction import TransactionStatus, TransactionType


class TransactionRead(BaseModel):
    id: UUID
    type: TransactionType
    status: TransactionStatus
    amount_paise: int
    currency: str
    description: str
    created_at: datetime


class TransactionListResponse(BaseModel):
    transactions: list[TransactionRead]

File: apps/web/lib/transactions.ts
// Transactions client for the authenticated dashboard.
//
// Calls /api/v1/transactions through the typed fetch wrapper in
// lib/api/client.ts. Wire shapes come from the generated contract; this maps
// them to the camelCase shape the UI consumes (same pattern as lib/site-visits.ts).
// Read-only: there is no create endpoint, rows come from a future money-layer producer.

import type { components } from "@contracts/generated/schema";

import { apiRequest, type ApiResponse } from "@/lib/api/client";

type Schemas = components["schemas"];

export type TransactionType = Schemas["TransactionType"];
export type TransactionStatus = Schemas["TransactionStatus"];

export type Transaction = {
  id: string;
  type: TransactionType;
  status: TransactionStatus;
  amountPaise: number;
  currency: string;
  description: string;
  createdAt: string;
};

function mapTransaction(raw: Schemas["TransactionRead"]): Transaction {
  return {
    id: raw.id,
    type: raw.type,
    status: raw.status,
    amountPaise: raw.amount_paise,
    currency: raw.currency,
    description: raw.description,
    createdAt: raw.created_at,
  };
}

export async function getTransactions(): Promise<ApiResponse<Transaction[]>> {
  const res = await apiRequest<Schemas["TransactionListResponse"]>("/api/v1/transactions");
  if (!res.ok) return res;
  return { ok: true, status: res.status, data: res.data.transactions.map(mapTransaction) };
}
